alibi_detect/cd/pytorch/context_aware.py

```diff
@@ -0,0 +1,264 @@
+        idx_0, idx_1 = np.where(bools == 0)[0], np.where(bools == 1)[0]
+        n_ref, n_test = len(idx_0), len(idx_1)
+
+        # Form kernel matrices
+        L_0, L_1 = L[idx_0][:, idx_0], L[idx_1][:, idx_1]
+        K_0, K_1 = K[idx_0][:, idx_0], K[idx_1][:, idx_1]
+
+        # Initialise regularisation parameters
+        # Implemented only for first _cmmd call which corresponds to original window assignment
+        if self.lams is None:
+            possible_lams = torch.tensor([2**(-i) for i in range(20)]).to(K.device)
+            lam_0 = self._pick_lam(possible_lams, K_0, L_0, n_folds=self.n_folds)
+            lam_1 = self._pick_lam(possible_lams, K_1, L_1, n_folds=self.n_folds)
+            self.lams = (lam_0, lam_1)
+
+        # Compute stat
+        L_0_inv = torch.linalg.inv(L_0 + n_ref*self.lams[0]*torch.eye(int(n_ref)).to(L_0.device))
+        L_1_inv = torch.linalg.inv(L_1 + n_test*self.lams[1]*torch.eye(int(n_test)).to(L_1.device))
+        A_0 = L_held[:, idx_0] @ L_0_inv
+        A_1 = L_held[:, idx_1] @ L_1_inv
+        # Allow batches of MMDs to be computed at a time (rather than all)
+        if self.batch_size is not None:
+            bs = self.batch_size
+            coupling_xx = torch.stack([torch.einsum('ij,ik->ijk', A_0_i, A_0_i).mean(0)
+                                       for A_0_i in A_0.split(bs)]).mean(0)
+            coupling_yy = torch.stack([torch.einsum('ij,ik->ijk', A_1_i, A_1_i).mean(0)
+                                       for A_1_i in A_1.split(bs)]).mean(0)
+            coupling_xy = torch.stack([
+                torch.einsum('ij,ik->ijk', A_0_i, A_1_i).mean(0) for A_0_i, A_1_i in zip(A_0.split(bs), A_1.split(bs))
+            ]).mean(0)
+        else:
+            coupling_xx = torch.einsum('ij,ik->ijk', A_0, A_0).mean(0)
+            coupling_yy = torch.einsum('ij,ik->ijk', A_1, A_1).mean(0)
+            coupling_xy = torch.einsum('ij,ik->ijk', A_0, A_1).mean(0)
+        sim_xx = (K[idx_0][:, idx_0]*coupling_xx).sum()
+        sim_yy = (K[idx_1][:, idx_1]*coupling_yy).sum()
+        sim_xy = (K[idx_0][:, idx_1]*coupling_xy).sum()
+        stat = sim_xx + sim_yy - 2*sim_xy
+
+        return stat.cpu(), coupling_xx.cpu(), coupling_yy.cpu(), coupling_xy.cpu()
+
+    def _pick_lam(self, lams: torch.Tensor, K: torch.Tensor, L: torch.Tensor, n_folds: int = 5) -> torch.Tensor:
+        """
+        The conditional mean embedding is estimated as the solution of a regularised regression problem.
+
+        This private method function uses cross validation to select the regularisation parameter that
+        minimises squared error on the out-of-fold instances. The error is a distance in the RKHS and is
+        therefore an MMD-like quantity itself.
+        """
+        n = len(L)
+        fold_size = n // n_folds
+        K, L = K.type(torch.float64), L.type(torch.float64)
+        losses = torch.zeros_like(lams, dtype=torch.float).to(K.device)
+        for fold in range(n_folds):
```

Comment:
A change I made to my private repo since sharing the prototype is to shuffle K and L prior to running the cross-val procedure. Would be important in cases where the user doesn't shuffle the data themselves and the ordering is non-random. I recommend adding
```
perm = torch.randperm(n)
K, L = K[perm][:,perm], L[perm][:,perm]
```

Reply:
Ah oops, I shuffled for the held out split but not for this. Will update!